Objects/frameobject.c

```diff
@@ -13,9 +13,320 @@
+    Py_TYPE(self)->tp_free(self);
+}
+
+int framelocalsproxy_init(PyObject *self, PyObject *args, PyObject *kwds)
+{
+    PyObject* frame = PyTuple_GET_ITEM(args, 0);
+    assert(PyFrame_Check(frame));
+    Py_INCREF(frame);
+    ((PyFrameLocalsProxyObject*)self)->frame = (PyFrameObject*)frame;
+
+    return 0;
+}
+
+PyObject* framelocalsproxy_iter(PyObject *self)
+{
+    return PyObject_GetIter(framelocalsproxy_keys(self, NULL));
+}
+
+PyObject* framelocalsproxy_richcompare(PyObject *self, PyObject *other, int op)
+{
+    if (PyFrameLocalsProxy_Check(other)) {
+        bool result = ((PyFrameLocalsProxyObject*)self)->frame == ((PyFrameLocalsProxyObject*)other)->frame;
+        if (op == Py_EQ) {
+            return PyBool_FromLong(result);
+        } else if (op == Py_NE) {
+            return PyBool_FromLong(!result);
+        }
+    } else if (PyDict_Check(other)) {
+        PyObject* dct = PyDict_New();
+        PyObject* result = NULL;
+        PyDict_Update(dct, self);
+        result = PyObject_RichCompare(dct, other, op);
+        Py_DECREF(dct);
+        return result;
+    }
+
+    Py_RETURN_NOTIMPLEMENTED;
+}
+
+// Methods
+
+PyObject* framelocalsproxy_items(PyObject* self, PyObject* __unused)
+{
+    PyObject* items = PyList_New(0);
+    _PyInterpreterFrame* frame = ((PyFrameLocalsProxyObject*)self)->frame->f_frame;
+    PyCodeObject* co = _PyFrame_GetCode(frame);
+
+    for (int i = 0; i < co->co_nlocalsplus; i++) {
+        PyObject* name = PyTuple_GET_ITEM(co->co_localsplusnames, i);
+        PyObject* value = framelocalproxy_getval(frame, co, i);
+
+        if (value) {
+            PyObject* pair = PyTuple_Pack(2, name, value);
+            PyList_Append(items, pair);
+            Py_DECREF(pair);
+        }
+    }
+
+    // Iterate through the extra locals
+    Py_ssize_t j = 0;
+    PyObject* key = NULL;
+    PyObject* value = NULL;
+
+    while (PyDict_Next(frame->f_extra_locals, &j, &key, &value)) {
+        PyObject* pair = PyTuple_Pack(2, key, value);
+        PyList_Append(items, pair);
+        Py_DECREF(pair);
+    }
+
+    return items;
+}
+
+Py_ssize_t framelocalsproxy_length(PyObject* self)
+{
+    _PyInterpreterFrame *frame = ((PyFrameLocalsProxyObject*)self)->frame->f_frame;
+    PyCodeObject *co = _PyFrame_GetCode(frame);
+    Py_ssize_t size = 0;
+    if (frame->f_extra_locals != NULL) {
+        assert(PyDict_Check(frame->f_extra_locals));
+        size += PyDict_Size(frame->f_extra_locals);
+    }
+
+    for (int i = 0; i < co->co_nlocalsplus; i++) {
+        if (framelocalproxy_getval(frame, co, i) != NULL) {
+            size++;
+        }
+    }
+    return size;
+}
+
+PyObject* framelocalsproxy_getitem(PyObject* self, PyObject* key)
+{
+    _PyInterpreterFrame *frame = ((PyFrameLocalsProxyObject*)self)->frame->f_frame;
+    PyCodeObject *co = _PyFrame_GetCode(frame);
+
+    for (int i = 0; i < co->co_nlocalsplus; i++) {
+        if (_PyUnicode_EQ(PyTuple_GET_ITEM(co->co_localsplusnames, i), key)) {
```

Comment:
The strings in `co_localsplusnames` will all be interned and `key` is likely to be interned, it is probably worth do a fast pass with `PyTuple_GET_ITEM(co->co_localsplusnames, i) == key` before falling back onto the full `_PyUnicode_EQ` equality check.

It might not worth it, though, as this is unlikely to be hot code.